src/cpp/client/client_context.cc

```diff
@@ -77,8 +77,12 @@ ClientContext::ClientContext()
 }
 
 ClientContext::~ClientContext() {
-  if (call_) {
-    grpc_call_unref(call_);
+  {
+    internal::MutexLock lock(&mu_);
```

Comment:
Is this lock required? Usually holding a lock in ctor or dtor isn't typical.

Reply:
I think this is required, as the test still crashes without holding the lock. Theoretically, someone can access `call_` after `grpc_call_unref()` but before `call_` being set to nullptr, which causes trouble.

Reply:
Accessing `client_context` while destroying it sounds wrong to me. Would you check why your test case crash without holding a lock? There could be a bug in your test code.

Reply:
It seems that the test code:
1. calls `ClientContext::TryCancel()` to speed up the test tear down, without calling `TryCancel()` the test still works but slower.
2. will not call `ClientContext::TryCancel()` if `ClientContext` has been "fully destructed", because the test code stops tracking `ClientContext` in its `GlobalCallbacks::Destructor()` (which `~ClientContext()` calls as its last statement).

IIUC, the race will be exposed when `TryCancel()` and `~ClientContext()` are called in different thread, to be more specific, when `~ClientContext()` was entered but `call_` hasn't been set to nullptr.

It looks to me that `ClientContext` is supposed to support multithreads environment and it should itself avoid such crash, or is the test code doing anything wrong here?

Reply:
I think the test code is wrong here. An object should not be used once its destructor is called, and it's the callers responsibility to ensure that.

Reply:
Setting `call_` to nullptr in the dtor of ClientContext would be reasonable to address the case where the global hook might want to access it in its callback but using the lock doesn't sound right in this case because we don't support case where a ClientContext instance is being destroyed and one of its methods is being called by different threads, which is a race condition.

Reply:
Thanks for replying!

@ctiller I do agree we shouldn't use the object once its destructor is called to some extent. But it is gRPC who offers the `GlobalCallbacks` class and its `Destructor()` function here for people to access the object even after its destructor is called (but not yet finished). The `Destructor()` function is given the `ClientContext` object as its argument. What kind of operation does gRPC expect the user to do in the `Destructor()` function if no one should use the object once its destructor is called? (See below paragraph for our use case :)

@veblush The test code is trying to adopt an easy way to "cancel all other calls if a particular `ClientContext` is going to be destroyed". The test code is not randomly calling `TryCancel()` in some random thread in this case, it keeps track of all `ClientContext` when `DefaultConstructor()` is called, and removes them in `Destructor()`. It carefully uses a mutex to ensure that if a `ClientContext` has been fully destroyed, `TryCancel()` will not be called. If this is not enough, then sounds like the `GlobalCallbacks` is not a good candidate to achieve such goal (cancel other calls when a particular `ClientContext` is going to be destroyed). Is there a suggested way to do that? Of course we can go all the way back to our context manager and add a function like "CancelAllCallsForTesting()`, but `GlobalCallbacks` looks like a good and simple shortcut that we really want to use :)

Reply:
We've removed our `GlobalCallbacks` which cancels all other RPC calls. It increases the testing time but it's acceptable. We don't need this anymore. The lock is removed.